image-registry/remove-registry-baremetal.md

```diff
@@ -0,0 +1,137 @@
+
+## Motivation
+
+IPI installation options for Kubernetes-native infrastructure (formerly RHHI.next) and Red Hat Virtualization (RHV) have been declared a key initiative for OpenShift 4.3.
+These will configure themselves as the `baremetal` platform type.
+
+In 4.2 the registry bootstraps itself to use `EmptyDir` storage on `baremetal` to enable pre-GA developemnt on these platforms ([PR #332](https://github.com/openshift/cluster-image-registry-operator/pull/332)).
+This is not suitable for production as there is a serious risk of data loss if a permanent form of storage is not configured.
+
+### Goals
+
+- `openshift-installer` completes if the cloud provider is set to `baremetal`.
+- Image registry is able to bring itself up after the management state is switched from `Removed` to `Managed`.
+
+### Non-Goals
+
+- Provision storage for the registry for RHV and k8s-native infrastructre.
+- Provide 1st class support for additional storage types/providers for the registry (ex: MCG).
+- Harden OpenShift features (ex: Builds, ImageStreams) such that they error cleanly if the registry is not installed.
+
+## Proposal
+
+This is where we get down to the nitty gritty of what the proposal actually is.
+
+### User Story
+
+As an OpenShift cluster admin/installer
+I want the registry to be `Removed` on RHV and k8s-native infra installations
+So that I can provision other storage providers
+And later enable the image registry with the desired storage
+
+### Implementation Details/Notes/Constraints
+
+- When the image registry operator bootstraps itself, it will inspect the cluster infrastructure platform status.
+- If the platform status type is `baremetal`, the registry will mark itself `Removed`.
+- A separate installation process will be resposible for provisioning storage. For RHV and k8s-native infra, this is expected to be a PVC with the `RWX` access mode.
+- Telemeter should be aware if a cluster's registry has been removed:
+  - Reason codes in the ClusterOperator conditions (`Available` in particular).
+  - Prometheus alerts warning registry has been removed.
+
+### Risks and Mitigations
+
+#### Image registry is never installed
+
+Several components will likely report errors if a customer runs the installer as `baremetal` and does nothing to set up the registry.
+Potential candidates that will break:
+
+1. Builds which push to an imagestream tag will error at runtime.
+2. Samples operator imagestreams which import from `registry.redhat.io` will require a pull secret to be usable.
+3. Imagestreams imported with `--reference-policy=local` will not reference the internal registry as expected. `BuildConfig` and `DeploymentConfig` objects which do their own resolution of `ImageStreamTag` will break.
+
+Mitigations:
+
+1. Existing error messages from affected components.
+2. Image registry's ClusterOperator will provide `Reason="Removed"` if the registry remains removed.
+3. If removed, the image registry operator should fire an alert warning that:
```

Comment:
As a brief reminder, we need to get [1] merged before we implement this one. 

[1] https://github.com/openshift/cluster-image-registry-operator/pull/368

Reply:
I remember I have seen somewhere that if we install a new cluster we can't have any other alert than the "test" one. How is this going to work? Are we going to accept the fact that in some cases we may have multiple alerts soon after installing the cluster?

Reply:
My gut is that our conformance suites will fail elsewhere unless there is an internal registry installed. So these platforms will need to automate day 2 activities as a part of their CI.